### PR TITLE
feat(db): add `zstd` and `CompactZstd` to `Transactions` and `Receipts`

### PR DESCRIPTION
**Pull request by joshieDo**
Adds [zstd](https://facebook.github.io/zstd/).

Dictionary trained on the first 2M blocks, for both `Transactions` and `Receipts`.  

## 5M:
main:
```
| Transactions      | 149860206 | 27516        | 6163331    | 410349         | 25.2 GiB   |
| Receipts          | 149860206 | 15809        | 3540941    | 404885         | 15.1 GiB   |
```
branch:
```
| Transactions      | 149860206 | 24737        | 5540827    | 110080         | 21.7 GiB   |
| Receipts          | 149860206 | 11769        | 2635875    | 46970          | 10.3 GiB   |
```

## Tip : 
main (17.14M): 

```
Transactions: 690GiB
Receipts: 501GiB
```

branch [17156161]:
```
Transactions: 406.4GiB
Receipts:  220GiB
```

## Discussion.

Tried to play with dictionary size, training data and compression data. A bigger dictionary doesn't necessarily mean better compression. But having one does provide it, alongside with improved speeds (check the zstd link). We might have better results if we tune these things (eg. train dictionaries from 0 to tip), but it just takes time...

Some questions:

1. The zstd execution branches on `from_compact` and `to_compact` require an extra buffer. Either to prepare the data to compress, or to decompress the data before decoding it to the type. In the decompress case, if the capacity is not enough it will throw an error, which we catch and then increase the buffer size. But given that this will have to occur millions of times during the first sync, I'm tempted to change the interface from: `from_compact(buf: &mut B)` to `from_compact(buf: &mut B, tmp: &mut B)`.  Same for `to_compact`.  It would allow us to re-use an extra buffer belonging to the `cursor`. As we already have for the original buf. What do you think?
2. Given that we go down this route. What would be an acceptable dictionary size? Currently they are 100K each, which get loaded into memory and go alongside the `reth` binary`.

### CODE REVIEW

**Comment by gakonst** [comment]
Epic. @joshieDo did you check what the decompression overhead looks like at runtime?

**Comment by gakonst** [comment]
bump

**Comment by gakonst** [comment]
Running this from genesis to block 12M to benchmark.

**Comment by joshieDo** [comment]
For the `Transactions` table a 50mb dictionary gives us 389GB vs 100K dictionary which gives us 406GB - is it worth to add the 50mb

**Review by gakonst — COMMENTED** [review]
Love the code & the perf outcome. Do you mind adding some more docs on how various parameters were chosen in the code, as well as explain a bit the global dictionary instantiation / why thread_local, and if the tmp buf that you used to have is no longer relevant?

File: crates/primitives/src/transaction/mod.rs [hunk 1]
```diff
@@ -773,19 +776,58 @@ impl Compact for TransactionSignedNoHash {
...
+                let mut decompressor = decompressor.borrow_mut();
+                let mut tmp: Vec<u8> = Vec::with_capacity(200);
+
+                while let Err(err) = decompressor.decompress_to_buffer(buf, &mut tmp) {
+                    let err = err.to_string();
+                    if !err.contains("Destination buffer is too small") {
+                        panic!("Failed to decompress: {}", err);
+                    }
+                    tmp.reserve(tmp.capacity() + 24_000);
+                }
```
[the remark this answers is not in the document]
Comment (joshieDo) [reply crates/primitives/src/transaction/mod.rs:824]: added docs, does it make sense?

File: crates/primitives/src/transaction/mod.rs [hunk 2]
```diff
@@ -773,19 +776,58 @@ impl Compact for TransactionSignedNoHash {
...
+
+                // TODO: enforce that zstd is only present at a "top" level type
+
+                let (transaction, _) =
+                    Transaction::from_compact(tmp.as_slice(), (prefix & 0b110) >> 1);
```
[the remark this answers is not in the document]
Comment (joshieDo) [reply crates/primitives/src/transaction/mod.rs]: changed a few var names + docs

**Review by gakonst — APPROVED** [review]
let's send it with the 100k dicts and we can revisit in the next big breakage

**Pull request merged**